:warning: Simplify the webhook interface

mengqiy opened the pull request:
If a CRD type implements the `Defaulter` and `Validator` interface, we will generate the admission webhook for the CRD type.
For non-CRD type, we provide methods to pass in path and handlers to create the mutating|validting webhook.

godoc and tests are on the way.

droot commented on pkg/webhook/server.go:
```diff
@@ -53,256 +52,145 @@ type ServerOptions struct {
+		if wh == nil {
+			continue
+		}
+		s.Register("path", wh)
```
We can generate the path using object's GVK and name. And, the default path generation should probably be shared with controller-tools so that generator is in sync with this code ?

droot commented on pkg/webhook/server.go:
```diff
@@ -53,256 +52,145 @@ type ServerOptions struct {
-	once sync.Once
+// For registers defaulting and validation webhooks for the provided types that implement either
+// webhooktuil.Validator or webhookutil.Defaulter.
+func (s *Server) For(objects ...runtime.Object) error {
```
`For` name itself doesn't convey the intent. Wondering if we should give it a more friendly name like `EnableValidationFor` or `EnableDefaultsFor` ?

mengqiy commented:
Addressed comments.
PTAL
If looks OK, I will push more tests and godoc.

droot commented on crdexample/pkg/resource.go:
```diff
@@ -0,0 +1,138 @@
+// TODO: figure out should we encourage the users to do a wrapper for a core type object.
+type MyPod struct {
+	corev1.Pod
+}
```
Its not clear to me in this case whether the webhook gets generated for the `corev1.Pods` (the embedded type) or the wrapped type ?

mengqiy replied:
This was some experimental code, but they actually sneak in.
I will remove this piece.
Sorry for the confusion.

droot commented on crdexample/pkg/zz_generated.deepcopy.go:
```diff
@@ -0,0 +1,127 @@
+	out := new(FirstMateStatus)
+	in.DeepCopyInto(out)
+	return out
+}
```
I have seen these example projects add too much noise to the repo and introduce additional maintenance burden. Is there a way to avoid it ?

droot commented on pkg/builder/build.go:
```diff
@@ -148,6 +148,18 @@ func (blder *Builder) Build(r reconcile.Reconciler) (manager.Manager, error) {
 		return nil, err
 	}
 
+	if blder.mgr.GetWebhookServer() != nil {
```
nit: we can combine the if conditional with the svr assignment.
```Go
 if svr := blder.mgr.GetWebhookServer(); svr != nil {
}
```

droot commented on pkg/builder/build.go:
```diff
@@ -148,6 +148,18 @@ func (blder *Builder) Build(r reconcile.Reconciler) (manager.Manager, error) {
+		if err != nil {
+			return nil, err
+		}
+		err = svr.EnableValidationFor(blder.mgr.GetScheme(), blder.apiType)
```
Won't `svr` have access to the scheme ? If yes, then we can get rid of scheme parameter from `EnableDefaultsFor` and `EnableValidationFor`.

mengqiy replied:
`svr` won't be able to have the Scheme until the inject time.
But the setup time is earlier than the inject time.

Another option is having 2 rounds of injection:
1st round injects deps for the server
2nd round inject deps to the handlers under the server.

It's not clear to me which way is better.

droot commented on pkg/manager/internal.go:
```diff
@@ -90,6 +91,8 @@ type controllerManager struct {
 
 	startCache func(stop <-chan struct{}) error
+
+	webhookServer *webhook.Server
```
is manager suppose to handle lifecycle of this webhookServer, then we should probably make it a `Runnable` ?

mengqiy replied:
`*webhook.Server` IS a `Runable`.

droot commented on pkg/webhook/handlers.go:
```diff
@@ -0,0 +1,190 @@
+)
+
+// Defaultor defines functions for setting defaults on resources
+type Defaulter interface {
```
typo: name in comment doesn't match.

droot commented on pkg/webhook/handlers.go:
```diff
@@ -0,0 +1,190 @@
+var _ inject.Scheme = &mutatingHandler{}
+
+// InjectDecoder injects the decoder into the FirstMateDeleteHandler
+func (h *mutatingHandler) InjectScheme(s *runtime.Scheme) error {
```
comment needs to be rewritten.

droot commented on pkg/webhook/handlers.go:
```diff
@@ -0,0 +1,190 @@
+//func (h *mutatingHandler) InjectDecoder(d *admission.Decoder) error {
+//	h.decoder = d
+//	return nil
+//}
```
we don't need this ?

droot commented on pkg/webhook/handlers.go:
```diff
@@ -0,0 +1,190 @@
+	if err != nil {
+		log.Error(err, "unable to get a decoder from the scheme")
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
```
I am sure we must have explored that but wanted to understand why this can't be done `once` for this handler at the start-up or may be at the time of scheme injection ?

droot commented on pkg/webhook/handlers.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+// newValidatingWebhookFor creates a new WebhookBuilder for Validation and Defaulting the provided type.  Returns nil
+// if o does not implement either validator or defaulter.
```
May be move the `validation` stuff in a separate file validation.go ?

droot commented on pkg/webhook/handlers.go:
```diff
@@ -0,0 +1,190 @@
+var _ inject.Scheme = &validatingHandler{}
+
+// InjectDecoder injects the decoder into the FirstMateDeleteHandler
+func (h *validatingHandler) InjectScheme(s *runtime.Scheme) error {
```
typo: comments need to be rewritten.

droot commented on pkg/webhook/handlers.go:
```diff
@@ -0,0 +1,190 @@
+	if err != nil {
+		log.Error(err, "unable to get a decoder from the scheme")
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
```
same comment as above.

droot commented on pkg/webhook/server.go:
```diff
@@ -82,6 +86,74 @@ func (s *Server) setDefaults() {
 
+// For registers defaulting and validation webhooks for the provided types that implement either
+// webhooktuil.validator or webhookutil.defaulter.
+func (s *Server) EnableDefaultsFor(scheme *runtime.Scheme, objects ...runtime.Object) error {
```
we can make scheme injected from manager and get rid of this parameter ?

mengqiy replied:
Same comment as https://github.com/kubernetes-sigs/controller-runtime/pull/328#discussion_r259543868

droot commented on pkg/webhook/server.go:
```diff
@@ -82,6 +86,74 @@ func (s *Server) setDefaults() {
+		if len(gvks) != 1 {
+			return fmt.Errorf("expected only GVK returned by scheme.ObjectKinds")
+		}
+		fmt.Println("/mutate-" + generatePath(gvks[0]))
```
logging this info at some verbosity level may not be a bad idea, can help in debugging.

mengqiy replied:
My debugging code sneaked in.
I agree it can be converted to logging.
Will fix it

droot commented on pkg/webhook/server.go:
```diff
@@ -82,6 +86,74 @@ func (s *Server) setDefaults() {
+			return err
+		}
+		if len(gvks) != 1 {
+			return fmt.Errorf("expected only GVK returned by scheme.ObjectKinds")
```
we should some webhook related context to the error message and also include all the GVKs to help in debugging.

droot commented on pkg/webhook/server.go:
```diff
@@ -82,6 +86,74 @@ func (s *Server) setDefaults() {
+
+// For registers defaulting and validation webhooks for the provided types that implement either
+// webhooktuil.validator or webhookutil.defaulter.
+func (s *Server) EnableValidationFor(scheme *runtime.Scheme, objects ...runtime.Object) error {
```
comment is out of sync due to renaming.

droot commented on pkg/webhook/server.go:
```diff
@@ -82,6 +86,74 @@ func (s *Server) setDefaults() {
+		s.Register("/validate-"+generatePath(gvks[0]), wh)
+	}
+	return nil
+}
```
I am sort of wondering if we should enable validation/defaulting by default when webhook server starts ? (This will save people one step if they want to enable validation/defaulting). And I am thinking of doing the same thing for conversion.

droot commented on pkg/webhook/server.go:
```diff
@@ -82,6 +86,74 @@ func (s *Server) setDefaults() {
+	}
+	// Register all of the types
+	s.Register(path, h)
+}
```
Name of the function doesn't make sense.

droot commented on pkg/webhook/server.go:
```diff
@@ -82,6 +86,74 @@ func (s *Server) setDefaults() {
+	}
+	// Register all of the types
+	s.Register(path, h)
+}
```
Name of the function doesn't make sense.

mengqiy commented:
Addressed comments. PTAL at the interfaces.
2 things are still open for discussion:
- should we do https://github.com/kubernetes-sigs/controller-runtime/pull/328#discussion_r259461431 
- should we have an example for CRD like the one in commit https://github.com/kubernetes-sigs/controller-runtime/pull/328/commits/b51eb44f9b3bca8a827817f71f94e221d793f1e3

@DirectXMan12 WDYT

droot commented on pkg/builder/build.go:
```diff
@@ -230,3 +234,22 @@ func (blder *Builder) doController(r reconcile.Reconciler) error {
+	// If we get a NotImplementDefaulterValidatorInterfacesError, we discord the Server instance.
+	// If other type of error, surface it.
+	if err == nil {
+		e := blder.mgr.Add(svr)
```
quick question: shouldn't this be done before calling `svr.EnableWebhooksFor` anyways to ensure mgr injects dependencies for webhooks (mainly scheme) ?

droot commented on pkg/webhook/defaulter.go:
```diff
@@ -0,0 +1,90 @@
+func (h *mutatingHandler) InjectScheme(s *runtime.Scheme) error {
+	h.scheme = s
+	var err error
+	h.decoder, err = admission.NewDecoder(h.scheme)
```
+1 on setting decoder this way.

DirectXMan12 commented on pkg/webhook/server.go:
```diff
@@ -74,14 +92,67 @@ func (s *Server) setDefaults() {
+// neither the Defaulter nor the Validator interfaces.
+// It returns a NotImplementDefaulterValidatorInterfacesError error if a provided object implements
+// neither the Defaulter interface nor the Validator interface.
+func (s *Server) EnableWebhooksFor(objects ...runtime.Object) error {
```
I'm not convinced this needs to be a method (as opposed to a helper or helpers).  The helper makes it more composable, easier to test individually, and lets us actually have a type signature that requires an object that implements the interfaces (removing the error path in favor of compile-time checking).

DirectXMan12 commented on pkg/webhook/server.go:
```diff
@@ -172,3 +243,15 @@ func (s *Server) InjectFunc(f inject.Func) error {
+
+func generatePath(gvk schema.GroupVersionKind) string {
+	var pathItems []string
+	splittedGroup := strings.Split(gvk.Group, ".")
```
why aren't we just `strings.Replace(gvk.Group, ".", "-")` or something?

mengqiy replied:
Curret behavior doesn't include domain name in the path:
```
Group: apps.example.com
Version: v1
Kind: Foo
```
I will get `/mutate-apps-v1-foo` instead of `/mutate-apps-example-com-v1-foo`. IMO it's more concise, but there may be a small risk of collision. e.g. with 
```
Group: apps.bar.com
Version: v1
Kind: Foo
```

I don't have a strong opinion about if we should have domain name.
Thoughts?

DirectXMan12 replied:
I don't think conciseness here is valuable (only ever have to write this once, and it's usually going to be autogenerated), and I'd rather avoid collisions

DirectXMan12 commented on pkg/webhook/validator.go:
```diff
@@ -0,0 +1,108 @@
+func (h *validatingHandler) InjectScheme(s *runtime.Scheme) error {
+	h.scheme = s
+	var err error
+	h.decoder, err = admission.NewDecoder(h.scheme)
```
decoders should automatically be injected with `InjectDecoder` by the `admission.Webhook` logic.

DirectXMan12 commented on pkg/builder/build.go:
```diff
@@ -218,3 +234,22 @@ func (blder *Builder) doController(r reconcile.Reconciler) error {
 }
+
+func (blder *Builder) doWebhook() error {
+	svr := &webhook.Server{}
```
or some other design -- but it needs to work for multiple controllers in a single binary

DirectXMan12 commented on pkg/webhook/server.go:
```diff
@@ -74,14 +92,67 @@ func (s *Server) setDefaults() {
+			return nil
+		}
+
+		mwh := newDefaultingWebhookFor(object)
```
we should opportunistically merge validating and defaulting hooks into a single hook when possible, to avoid the extra round-trip

mengqiy replied:
> we should opportunistically merge validating and defaulting hooks into a single hook when possible, to avoid the extra round-trip

This means you want to put defaulting and validating hooks in one mutatingWebhook. But doing it this way is not safe, since there may be another mutating webhook changes the object after your validation logic. The safe way to do it is making validation logic in a validating webhook, which is always run after all mutating webhooks.

DirectXMan12 replied:
good point :+1:

mengqiy commented:
`pkg` dir is ready for another round of reviewing interfaces.
I haven't make a lot of changes in `examples/crd`. I'd like to discuss with @DirectXMan12 more before spending more than time on it.

DirectXMan12 commented on pkg/webhook/server.go:
```diff
@@ -54,6 +58,9 @@ type Server struct {
 	CertDir string
 
+	// Scheme knows the mapping between go struct types and GVKs.
+	Scheme *runtime.Scheme
```
what's the scheme used here for?

DirectXMan12 replied:
(the comment doesn't actually say)

DirectXMan12 commented on pkg/webhook/admission/defaulter.go:
```diff
@@ -0,0 +1,85 @@
+var _ inject.Scheme = &mutatingHandler{}
+
+// InjectScheme injects the scheme into a mutatingHandler.
+func (h *mutatingHandler) InjectScheme(s *runtime.Scheme) error {
```
what's the scheme here for?

DirectXMan12 commented on examples/crd/main.go:
```diff
@@ -0,0 +1,67 @@
+
+	appsv1 "k8s.io/api/apps/v1"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	"sigs.k8s.io/controller-runtime/examples/crd/pkg"
```
let's use `ctrl "sigs.k8s.io/controller-runtime"` instead of most of these in the examples

DirectXMan12 commented on examples/crd/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+# Build the manager binary
```
don't think we should have the deployment stuff here -- this should be only code.  Deployment examples can live in KB

mengqiy replied:
Ack

DirectXMan12 commented on examples/crd/pkg/resource.go:
```diff
@@ -0,0 +1,118 @@
+var _ admission.Validator = &FirstMate{}
+
+// ValidateCreate implements webhookutil.validator so a webhook will be registered for the type
+func (f *FirstMate) ValidateCreate() error {
```
should have webhook generation annotations?

mengqiy commented on pkg/webhook/admission/validator.go:
```diff
@@ -0,0 +1,103 @@
+
+// Handle handles admission requests.
+func (h *validatingHandler) Handle(ctx context.Context, req Request) Response {
+	if h.validator == nil {
```
Done

mengqiy commented on pkg/builder/build.go:
```diff
@@ -218,3 +240,60 @@ func (blder *Builder) doController(r reconcile.Reconciler) error {
 }
+
+func (blder *Builder) doWebhook() error {
+	if blder.mgr.GetScheme() == nil {
```
Was being more defensive here. But if you think it's not necessary at all, I will drop it.

mengqiy commented:
PTAL

mengqiy commented:
Rebased on HEAD to resolved the conflicts and squashed some commits.
Ready for the final round of review

DirectXMan12 commented on pkg/builder/build.go:
```diff
@@ -218,3 +236,36 @@ func (blder *Builder) doController(r reconcile.Reconciler) error {
+		gvk.Version + "-" + strings.ToLower(gvk.Kind)
+
+	defaulter, isDefaulter := blder.apiType.(admission.Defaulter)
+	if isDefaulter {
```
nit: collapse this into a compound statement.

DirectXMan12 commented on pkg/builder/build.go:
```diff
@@ -218,3 +236,36 @@ func (blder *Builder) doController(r reconcile.Reconciler) error {
+		}
+	}
+
+	validator, isValidator := blder.apiType.(admission.Validator)
```
nit: collapse this into a compound statement

DirectXMan12 commented on pkg/builder/build.go:
```diff
@@ -218,3 +236,36 @@ func (blder *Builder) doController(r reconcile.Reconciler) error {
+		vwh := admission.ValidatingWebhookFor(validator)
+		if vwh != nil {
+			path := "/validate-" + partialPath
+			log.Info("registering a validating webhook", "path", path)
```
log the GVK?

DirectXMan12 commented on pkg/builder/build.go:
```diff
@@ -218,3 +236,36 @@ func (blder *Builder) doController(r reconcile.Reconciler) error {
+
+	validator, isValidator := blder.apiType.(admission.Validator)
+	if isValidator {
+		vwh := admission.ValidatingWebhookFor(validator)
```
when we land conversion, we're going to want to automatically handle all registered versions of a given group-kind.  please add a comment.

DirectXMan12 commented:
/lgtm
/approve

Can you follow up with a PR with slightly more comprehensive tests?

mengqiy commented:
> Can you follow up with a PR with slightly more comprehensive tests?

Sure!

mengqiy commented:
Forgot to remove the `do-not-merge` label last Friday, and this PR ended up with a small merge conflict with #352 in the comments.
I rebased and resolved conflicts.
